README.md
# Introduction to es-printf #

es-printf is an implementation of the standard C printf function targetted
at small embedded systems with very restricted code and data space.
It provides several advantages over the standard library function:

1. There are many fine-grained customisations which allow you to trim
the size to absolute minimum so it will fit in limited code space.
You can elect to have only hexadecimal output, full floating point support
and any number of options in between. The smallest configuration typically
uses less than 200 bytes of code space.

1. It does not use any static buffers so it is thread-safe and can
safely be used within interrupts and in an RTOS environment.

1. You provide your own output function which can be to a serial port
or to anything else convenient.

1. Because the source code is available it is easy to extend the
formatter to provide custom specifiers for your particular needs.

Full documentation is provided in [es-printf.html](docs/es-printf.html)


Changes.txt
Changelog for es-printf
-----------------------

This is just a summary of the main changes in each released version.
For detailed changes please check version history in the repository.

Version 1.1 - 01 Aug 2015
* Fixed several formatting bugs in floating point.
* Added support for leading zeros in floating point output.
* Added more test cases.
* Improved test program highlights any differences from the standard printf.

Version 1.0 - 26 Jun 2015
First public release on Sourceforge. Code previously residing on googlecode
had never been formally released.

The original basic integer version had been in use for over 10 years when
es-printf was started. It lacked many formatting options and floating point
support, and it was not configurable. es-printf was developed to offer
the full functionality of the standard printf function, but with full
configurability, optimisation for small micros, and thread-safe operation.


src/test.c
/*************************************************************************
es-printf  -  configurable printf for embedded systems

Test program to compare output with the standard library function
and to generate AVR code for embedded testing.

$Id$

**************************************************************************
Copyright (c) 2006 - 2013 Skirrid Systems

Permission is hereby granted, free of charge, to any person obtaining a
copy of this software and associated documentation files (the "Software"),
to deal in the Software without restriction, including without limitation
the rights to use, copy, modify, merge, publish, distribute, sublicense,
and/or sell copies of the Software, and to permit persons to whom the
Software is furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
DEALINGS IN THE SOFTWARE.
******************************************************************************/

#ifdef TEST_AVR
    #include "printf.h"
    #include <avr/io.h>
    #include <avr/pgmspace.h>
    #define GET_FORMAT(p)   pgm_read_byte(p)
    void outchar(char c) {}
    #define PUTCHAR_FUNC    outchar
    #define NO_DOUBLE_PRECISION
#else
    #include <stdio.h>
    #define EXP_3_DIGIT
#endif
    #include "float.h"

/* Force printf.c to rename the functions so we can use both
   the standard library and our own version for side-by-side tests.
*/
#define printf x
#define sprintf x

/* Include the printf source here so we can control the definitions
   used in the test environment. You would not normally need to do this.
*/
#include "printf.c"

#ifdef TEST_AVR
    /* In the AVR test environment define a new macro to use our function
       with strings in flash.
    */
    #define tprintf(format, args...)        printf_rom(PSTR(format), ## args)
  #ifdef BASIC_PRINTF_ONLY
    #define tsprintf(buf, format, args...)  printf_rom(PSTR(format), ## args)
  #else
    #define tsprintf(buf, format, args...)  sprintf_rom(buf, PSTR(format), ## args)
  #endif
#else
    /* In the PC test environment reinstate printf to call the normal library function.
       Define a new macro which calls both the library and our own versions of printf.
    */
    #undef printf
    #undef sprintf
    #define tprintf(format, args...)        do { printf(format, ## args); printf_rom(format, ## args); } while(0)
  #ifdef BASIC_PRINTF_ONLY
    #define tsprintf(buf, format, args...)  do { printf(format, ## args); printf_rom(format, ## args); } while(0)
  #else
    #define tsprintf(buf, format, args...)  do { printf(format, ## args); sprintf_rom(buf, format, ## args); printf("%s", buf); } while(0)
  #endif
#endif


#define PI  3.14159265358979323844
#define F6  123450.0
#define FL  1.2345e10
#define FS  1.2345e-10
#define N   -23
#define X   0x5a
#define O   023
#define S   "Abcde"

int main(int argc, char *argv[])
{
#ifdef SIZE
    #ifndef SIZEN
    tprintf("");
    #endif
#else
#ifndef BASIC_PRINTF_ONLY
    char buf[32];
#endif
#if FEATURE(USE_FLOAT)
    double one = 1.0;
#endif
    tsprintf(buf, "Hello world %d %% %z\n", 123);
#if FEATURE(USE_SPACE_PAD) || FEATURE(USE_ZERO_PAD)
    tprintf("Int %d %4d %04d %+04d % 4d %-4d. %+3d %+3d % 3d % 3d\n", N, N, N, N, N, N, 0, -1, 0, -1);
  #if FEATURE(USE_HEX_UPPER) || FEATURE(USE_HEX_LOWER)
    tprintf("Hex %x %4x %018X %+x % x %-4x. %+#x\n", X, X, X, X, X, X, X);
  #endif
  #if FEATURE(USE_OCTAL)
    tprintf("Oct %o %4o %04o %+o % o %-4o. %+#o\n", O, O, O, O, O, O, O);
  #endif
    tprintf("Str [%s] [%8s] [%-8s]\n", S, S, S);
  #if FEATURE(USE_INDIRECT)
    tprintf("Ind %*d %*s.\n", 4, N, 8, S);
  #endif
  #if FEATURE(USE_PRECISION)
    tprintf("Prec %.d %.2d %.2d %.2d %-5.3s %.3s.\n", 0, 1, 0, -1, S, S);
  #endif
#else
    tprintf("Int %d %+d %+d % d % d\n", N, 0, -1, 0, -1);
  #if FEATURE(USE_HEX_UPPER) || FEATURE(USE_HEX_LOWER)
    tprintf("Hex %x %+x % x\n", X, X, X);
  #endif
  #if FEATURE(USE_OCTAL)
    tprintf("Oct %o %+o % o\n", O, O, O);
  #endif
    tprintf("Str [%s]\n", S);
  #if FEATURE(USE_PRECISION)
    tprintf("Prec %.d %.2d %.2d %.2d %.3s.\n", 0, 1, 0, -1, S);
  #endif
#endif
#if FEATURE(USE_FLOAT)
    tprintf("pi = %f %e %g\n", PI, PI, PI);
    tprintf("f6dig = %.0f %e %g\n", F6, F6, F6);
    tprintf("flarge = %f %e %g\n", FL, FL, FL);
    tprintf("fsmall = %f %e %g\n", FS, FS, FS);
    tprintf("%%g = %g %g %g %G\n", 1.23e-5, 1.23e-4, 1.23e5, 1.23e6);
    tprintf("Max/Min = %e %e\n", DBL_MAX, DBL_MIN);
    tprintf("NaN/Inf = %f %f\n", sqrt(-1), one / (one - 1.0));
#endif
#endif
    return 0;
}


test/test.c
/*************************************************************************
es-printf  -  configurable printf for embedded systems

Test program to compare output with the standard library function
and to generate AVR code for embedded testing.

$Id$

**************************************************************************
Copyright (c) 2006 - 2015 Skirrid Systems

Permission is hereby granted, free of charge, to any person obtaining a
copy of this software and associated documentation files (the "Software"),
to deal in the Software without restriction, including without limitation
the rights to use, copy, modify, merge, publish, distribute, sublicense,
and/or sell copies of the Software, and to permit persons to whom the
Software is furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
DEALINGS IN THE SOFTWARE.
******************************************************************************/

#if defined(TEST_AVR)
    // 8-bit AVR tests
    #include "../src/printf.h"
    #include <avr/io.h>
    #include <avr/pgmspace.h>
    #define GET_FORMAT(p)   pgm_read_byte(p)
  #ifdef SIZE
    void outchar(char c) {}
  #else
    // Real output functions for ATMega running at 8MHz.
    #define F_CPU       8000000
    #define BAUD_RATE   19200
    #define BAUD_RELOAD     (F_CPU / 8 / BAUD_RATE - 1)
    #include <util/delay.h>
    void outchar(char c)
    {
        while (!(UCSRA & (1<<UDRE)))
            ;
        UDR = c;
    }
    void outinit(void)
    {
        UCSRA = (1<<U2X);
        UCSRB = (1<<TXEN);
        UCSRC = (3<<UCSZ0);     // Async, no parity, 1 stop, 8 data
        UBRRH = BAUD_RELOAD >> 8;
        UBRRL = BAUD_RELOAD & 0xff;
        _delay_ms(100);
        outchar('\n');
    }
  #endif
    #define PUTCHAR_FUNC    outchar
    #define NO_DOUBLE_PRECISION
#elif defined(TEST_STM8)
    // STM8 tests
    #include "../src/printf.h"
    void outchar(char c) { (void) c; }
    #define PUTCHAR_FUNC    outchar
    #define NO_DOUBLE_PRECISION
    #define NO_ISNAN_ISINF
#else
    // PC tests
    #include <stdio.h>
    // The PC library version always produces 3-digit exponents.
    // Force the same thing in our code to make comparison easier.
    #define EXP_3_DIGIT
    // Redefine our printf output function.
    static void testchar(char c);
    #define PUTCHAR_FUNC    testchar
#endif
    #include "float.h"

// The next section pulls in the real printf code.
// Exclude this for the test version with no printf included.
#if !(defined(SIZE) && defined(SIZEN))

/* Force printf.c to rename the functions so we can use both
   the standard library and our own version for side-by-side tests.
*/
#define printf x
#define sprintf x

/* Include the printf source here so we can control the definitions
   used in the test environment. You would not normally need to do this.
*/
#include "../src/printf.c"

#if defined(TEST_AVR)
    /* In the AVR test environment define a new macro to use our function
       with strings in flash.
    */
    #define tprintf(format, args...)        printf_rom(PSTR(format), ## args)
  #ifdef BASIC_PRINTF_ONLY
    #define tsprintf(buf, format, args...)  printf_rom(PSTR(format), ## args)
  #else
    #define tsprintf(buf, format, args...)  sprintf_rom(buf, PSTR(format), ## args); printf_rom(PSTR("%s"), buf)
  #endif
#elif defined(TEST_STM8)
    /* In the STM8 test environment the macros only have to call our functions.
       The Raisonance compiler does not support the ## args extension so revert
       to the less flexible __VA_ARGS__ instead.
    */
    #define tprintf(...)                    printf_rom(__VA_ARGS__)
  #ifdef BASIC_PRINTF_ONLY
    #define tsprintf(buf, ...)              printf_rom(__VA_ARGS__)
  #else
    #define tsprintf(buf, ...)              sprintf_rom(buf, __VA_ARGS__)
  #endif
#else
    /* In the PC test environment reinstate printf to call the normal library function.
       Define a new macro which calls both the library and our own versions of printf.
    */
    #undef printf
    #undef sprintf
    #define COMPARE_TEST
    #define tprintf(format, args...)        do { sprintf(stdbuf, format, ## args);  \
                                                 testinit();                        \
                                                 printf_rom(format, ## args);       \
                                                 testcompare(); } while(0)
  #ifdef BASIC_PRINTF_ONLY
    #define tsprintf(buf, format, args...)  do { sprintf(stdbuf, format, ## args);  \
                                                 testinit();                        \
                                                 printf_rom(format, ## args);       \
                                                 testcompare(); } while(0)
  #else
    #define tsprintf(buf, format, args...)  do { sprintf(stdbuf, format, ## args);  \
                                                 sprintf_rom(testbuf, format, ## args); \
                                                 testcompare(); } while(0)
  #endif
#endif

#endif  // !(defined(SIZE) && defined(SIZEN))

#ifdef SIZE
/* This variant is used only to determine the compiled size.
   It calls the printf function with an empty format string.
   The SIZEN option builds the same output file but without
   the printf function. By subtracting the two values we can
   arrive at a compiled output size including any library
   functions which are required.
*/
int main(int argc, char *argv[])
{
  #ifndef SIZEN
    tprintf("");
  #endif
    (void) argc;    // Suppress compiler warning about unused arguments.
    (void) argv;
    return 0;
}
#else   // ifdef SIZE
/* This variant is used to test the output with a variety of
   different format strings and output types.
   The PC version is used to check that the output of our printf
   function matches the output of the standard library function.
   The other versions are used to generate real embedded versions
   so the output can be tested on a live system. The latter requires
   a real output function to be provided.
*/

#define PI  3.14159265358979323844
#define F6  123450.0
#define FL  1.2345e10
#define FS  1.2345e-10
#define N   -23
#define X   0x5a
#define O   023
#define S   "Abcde"

#ifdef COMPARE_TEST
static char stdbuf[80];
static char testbuf[80];
static int testindex;

void testinit(void)
{
    testindex = 0;
    testbuf[0] = '\0';
}
void testchar(char c)
{
    if (testindex < sizeof(testbuf)-1)
    {
        testbuf[testindex++] = c;
        testbuf[testindex] = '\0';
    }
}
void testcompare(void)
{
    if (strcmp(stdbuf, testbuf) == 0)
    {
        printf("     %s", stdbuf);
    }
    else
    {
        printf("Std  %s", stdbuf);
        printf("Test %s", testbuf);
    }
}
#endif

int main(int argc, char *argv[])
{
#ifndef BASIC_PRINTF_ONLY
    char buf[32];
#endif
#if FEATURE(USE_FLOAT)
    double one = 1.0;
#endif

#if defined(TEST_AVR)
    outinit();
#endif

#ifdef COMPARE_TEST
    // Show the output order
    printf("Output is compared to the standard printf output.\n");
    printf("For matching results only one line is shown.\n");
    printf("If there is a difference, both are shown.\n");
#endif

    // Test sprintf function.
    tsprintf(buf, "Hello world %x %% %z\n", 0x123);

// Integer output
#if FEATURE(USE_SPACE_PAD) || FEATURE(USE_ZERO_PAD)
  // Tests for variants which include the padding options.
  #if FEATURE(USE_SIGNED)
    tprintf("Int %d %4d %04d %+04d % 4d %-4d. %+3d %+3d % 3d % 3d\n", N, N, N, N, N, N, 0, -1, 0, -1);
  #endif
  #if FEATURE(USE_HEX_UPPER) || FEATURE(USE_HEX_LOWER)
    tprintf("Hex %x %4x %018X %+x % x %-4x. %+#x\n", X, X, X, X, X, X, X);
  #endif
  #if FEATURE(USE_OCTAL)
    tprintf("Oct %o %4o %04o %+o % o %-4o. %+#o\n", O, O, O, O, O, O, O);
  #endif
  #if FEATURE(USE_BINARY)
    tprintf("Bin %b %4b %04b %+b % b %-4b. %+#b\n", O, O, O, O, O, O, O);
  #endif
  #if FEATURE(USE_STRING)
    tprintf("Str [%s] [%8s] [%-8s]\n", S, S, S);
  #endif
  #if FEATURE(USE_INDIRECT)
    tprintf("Ind %*d %*s.\n", 4, N, 8, S);
  #endif
  #if FEATURE(USE_PRECISION)
    tprintf("Prec %.d %.2d %.2d %.2d %-5.3s %.3s.\n", 0, 1, 0, -1, S, S);
  #endif
#else
  // Simpler tests for variants with no padding options.
  #if FEATURE(USE_SIGNED)
    tprintf("Int %d %+d %+d % d % d\n", N, 0, -1, 0, -1);
  #endif
  #if FEATURE(USE_HEX_UPPER) || FEATURE(USE_HEX_LOWER)
    tprintf("Hex %x %+x % x\n", X, X, X);
  #endif
  #if FEATURE(USE_OCTAL)
    tprintf("Oct %o %+o % o\n", O, O, O);
  #endif
    tprintf("Str [%s]\n", S);
  #if FEATURE(USE_PRECISION)
    tprintf("Prec %.d %.2d %.2d %.2d %.3s.\n", 0, 1, 0, -1, S);
  #endif
#endif

// Floating point output.
#if FEATURE(USE_FLOAT)
    tprintf("pi = %f %e %g\n", PI, PI, PI);
    tprintf("f6dig = %.0f %e %g\n", F6, F6, F6);
    tprintf("flarge = %f %e %g\n", FL, FL, FL);
    tprintf("fsmall = %f %e %g\n", FS, FS, FS);
    tprintf("%%g = %g %g %g %G\n", 1.23e-5, 1.23e-4, 1.23e5, 1.23e6);
    tprintf("Max/Min = %e %e\n", DBL_MAX, DBL_MIN);
    tprintf("NaN/Inf = %f %f\n", sqrt(-1), one / (one - 1.0));
    tprintf("whole = %.f %.e %.g\n", 27.0, 27.0, 27.0);
    tprintf("pi-p4 = %.4f %.4e %.4g\n", PI, PI, PI);
    tprintf("one-p0 = %.f %.e %.g\n", 1.0, 1.0, 1.0);
    tprintf("one#-p0 = %#.f %#.e %#.g\n", 1.0, 1.0, 1.0);
    tprintf("zero = %f %e %g\n", 0.0, 0.0, 0.0);
    tprintf("zero-p0 = %.f %.e %.g\n", 0.0, 0.0, 0.0);
    tprintf("zero-p1 = %.1f %.1e %.1g\n", 0.0, 0.0, 0.0);
#endif

// String-in-flash output, only relevant to AVR.
#if FEATURE(USE_FSTRING) && defined(TEST_AVR)
    tprintf("FStr [%s] [%S] [%8S] [%-8S]\n", S, PSTR(S), PSTR(S), PSTR(S));
#endif

    (void) argc;    // Suppress compiler warning about unused arguments.
    (void) argv;

#if defined(TEST_AVR)
    // Never return from embedded test version.
    for (;;) ;
#endif

    return 0;
}
#endif  // ifdef SIZE


trunk/src/test.c
/*************************************************************************
es-printf  -  configurable printf for embedded systems

Test program to compare output with the standard library function
and to generate AVR code for embedded testing.

$Id$

**************************************************************************
Copyright (c) 2006 - 2013 Skirrid Systems

Permission is hereby granted, free of charge, to any person obtaining a
copy of this software and associated documentation files (the "Software"),
to deal in the Software without restriction, including without limitation
the rights to use, copy, modify, merge, publish, distribute, sublicense,
and/or sell copies of the Software, and to permit persons to whom the
Software is furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
DEALINGS IN THE SOFTWARE.
******************************************************************************/

#ifdef TEST_AVR
    #include "printf.h"
    #include <avr/io.h>
    #include <avr/pgmspace.h>
    #define GET_FORMAT(p)   pgm_read_byte(p)
    void outchar(char c) {}
    #define PUTCHAR_FUNC    outchar
    #define NO_DOUBLE_PRECISION
#else
    #include <stdio.h>
    #define EXP_3_DIGIT
#endif
    #include "float.h"

/* Force printf.c to rename the functions so we can use both
   the standard library and our own version for side-by-side tests.
*/
#define printf x
#define sprintf x

/* Include the printf source here so we can control the definitions
   used in the test environment. You would not normally need to do this.
*/
#include "printf.c"

#ifdef TEST_AVR
    /* In the AVR test environment define a new macro to use our function
       with strings in flash.
    */
    #define tprintf(format, args...)        printf_rom(PSTR(format), ## args)
  #ifdef BASIC_PRINTF_ONLY
    #define tsprintf(buf, format, args...)  printf_rom(PSTR(format), ## args)
  #else
    #define tsprintf(buf, format, args...)  sprintf_rom(buf, PSTR(format), ## args)
  #endif
#else
    /* In the PC test environment reinstate printf to call the normal library function.
       Define a new macro which calls both the library and our own versions of printf.
    */
    #undef printf
    #undef sprintf
    #define tprintf(format, args...)        do { printf(format, ## args); printf_rom(format, ## args); } while(0)
  #ifdef BASIC_PRINTF_ONLY
    #define tsprintf(buf, format, args...)  do { printf(format, ## args); printf_rom(format, ## args); } while(0)
  #else
    #define tsprintf(buf, format, args...)  do { printf(format, ## args); sprintf_rom(buf, format, ## args); printf("%s", buf); } while(0)
  #endif
#endif


#define PI  3.14159265358979323844
#define F6  123450.0
#define FL  1.2345e10
#define FS  1.2345e-10
#define N   -23
#define X   0x5a
#define O   023
#define S   "Abcde"

int main(int argc, char *argv[])
{
#ifndef BASIC_PRINTF_ONLY
    char buf[32];
#endif
#ifdef SIZE
    tsprintf(buf, "Size %d\n", 123);
    tprintf("Size %d\n", 123);
#else
#if FEATURE(USE_FLOAT)
    double one = 1.0;
#endif
    tsprintf(buf, "Hello world %d %% %z\n", 123);
#if FEATURE(USE_SPACE_PAD) || FEATURE(USE_ZERO_PAD)
    tprintf("Int %d %4d %04d %+04d % 4d %-4d. %+3d %+3d % 3d % 3d\n", N, N, N, N, N, N, 0, -1, 0, -1);
  #if FEATURE(USE_HEX_UPPER) || FEATURE(USE_HEX_LOWER)
    tprintf("Hex %x %4x %018X %+x % x %-4x. %+#x\n", X, X, X, X, X, X, X);
  #endif
  #if FEATURE(USE_OCTAL)
    tprintf("Oct %o %4o %04o %+o % o %-4o. %+#o\n", O, O, O, O, O, O, O);
  #endif
    tprintf("Str [%s] [%8s] [%-8s]\n", S, S, S);
  #if FEATURE(USE_INDIRECT)
    tprintf("Ind %*d %*s.\n", 4, N, 8, S);
  #endif
  #if FEATURE(USE_PRECISION)
    tprintf("Prec %.d %.2d %.2d %.2d %-5.3s %.3s.\n", 0, 1, 0, -1, S, S);
  #endif
#else
    tprintf("Int %d %+d %+d % d % d\n", N, 0, -1, 0, -1);
  #if FEATURE(USE_HEX_UPPER) || FEATURE(USE_HEX_LOWER)
    tprintf("Hex %x %+x % x\n", X, X, X);
  #endif
  #if FEATURE(USE_OCTAL)
    tprintf("Oct %o %+o % o\n", O, O, O);
  #endif
    tprintf("Str [%s]\n", S);
  #if FEATURE(USE_PRECISION)
    tprintf("Prec %.d %.2d %.2d %.2d %.3s.\n", 0, 1, 0, -1, S);
  #endif
#endif
#if FEATURE(USE_FLOAT)
    tprintf("pi = %f %e %g\n", PI, PI, PI);
    tprintf("f6dig = %.0f %e %g\n", F6, F6, F6);
    tprintf("flarge = %f %e %g\n", FL, FL, FL);
    tprintf("fsmall = %f %e %g\n", FS, FS, FS);
    tprintf("%%g = %g %g %g %G\n", 1.23e-5, 1.23e-4, 1.23e5, 1.23e6);
    tprintf("Max/Min = %e %e\n", DBL_MAX, DBL_MIN);
    tprintf("NaN/Inf = %f %f\n", sqrt(-1), one / (one - 1.0));
#endif
#endif
    return 0;
}


tags/es-printf-1.0/src/printf.h
/*************************************************************************
es-printf  -  configurable printf for embedded systems

printf.h: Header for consumers of printf functions.

$Id$

**************************************************************************
Copyright (c) 2006 - 2015 Skirrid Systems

Permission is hereby granted, free of charge, to any person obtaining a
copy of this software and associated documentation files (the "Software"),
to deal in the Software without restriction, including without limitation
the rights to use, copy, modify, merge, publish, distribute, sublicense,
and/or sell copies of the Software, and to permit persons to whom the
Software is furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
DEALINGS IN THE SOFTWARE.
******************************************************************************/

#ifndef PRINTF_H
#define PRINTF_H

/*************************************************************************
Number of chars output

Traditionally printf returns the number of chars output. If you are not
interested in that value you can leave PRINTF_T undefined.
On a small micro you can define the return type as unsigned char if you
are sure the total output width will never exceed 255, or unsigned short.
*************************************************************************/

#define PRINTF_T unsigned short

// Create a type definition for the return value
#ifndef PRINTF_T
typedef void printf_t;
#else
typedef PRINTF_T printf_t;
#endif

/*************************************************************************
Memory access definitions

Some micros such as the AVR can only support storing strings in flash
memory by wrapping the string in a macro. To make this transparent we can
define the printf function itself as a macro which performs the wrap and
calls a renamed version of printf with a _rom suffix.
*************************************************************************/

/* Example for AVR micros using WinAVR (GCC) compiler

#define sprintf(buf, format, args...)   sprintf_rom(buf, PSTR(format), ## args)
#define printf(format, args...)         printf_rom(PSTR(format), ## args)

extern printf_t sprintf_rom(char *, const char *, ...);
extern printf_t printf_rom(const char *, ...);
*/

/*************************************************************************
Standard declarations

These are the declarations for the standard functions, unless overridden
by the memory access macros above.
*************************************************************************/

#ifndef printf
extern printf_t printf(const char *, ...);
#endif

#ifndef sprintf
extern printf_t sprintf(char *, const char *, ...);
#endif

#endif


tags/es-printf-1.0/src/printf_cfg.h
/*************************************************************************
es-printf  -  configurable printf for embedded systems

print_cfg.h: Select the build options for printf functions.

$Id$

**************************************************************************
Copyright (c) 2006 - 2015 Skirrid Systems

Permission is hereby granted, free of charge, to any person obtaining a
copy of this software and associated documentation files (the "Software"),
to deal in the Software without restriction, including without limitation
the rights to use, copy, modify, merge, publish, distribute, sublicense,
and/or sell copies of the Software, and to permit persons to whom the
Software is furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
DEALINGS IN THE SOFTWARE.
******************************************************************************/

#ifndef PRINTF_CFG_H
#define PRINTF_CFG_H

/*************************************************************************
Basic printf only

The code is designed to support a variety of printf-related functions.
If simple serial output is all you want then you can save some space by
defining BASIC_PRINTF_ONLY which allows the internal API to be simplified.
Note that sprintf will not be supported in this case.
*************************************************************************/

//#define BASIC_PRINTF_ONLY

/*************************************************************************
Memory access definitions

Some micros such as the AVR can only support storing and accessing strings
in flash memory using special macros and functions. This section can be
used to specify those methods. You may also need to modify printf.h
to get the compiler to place the format strings in flash memory.

The GET_FORMAT(ptr) macro is used to access a character in the printf
format string. By default this does a normal memory pointer access, but
you can configure it to access flash memory if needed.
*************************************************************************/

/* Example for AVR micros using WinAVR (GCC) compiler

#include <avr/io.h>
#include <avr/pgmspace.h>
#define GET_FORMAT(p)   pgm_read_byte(p)
*/

/*************************************************************************
Output configuration

By default printf will use the putchar function. If this is not defined
in your system you can set your own function here by defining
PUTCHAR_FUNC to be the name of that function.
*************************************************************************/

/*
extern void UARTSendByte(char c);
#define PUTCHAR_FUNC    UARTSendByte
*/

/*************************************************************************
Compiler capability configuration

Set some options that the C pre-processor will not tell us about.
*************************************************************************/

// Does the compiler support double precision or silently degrade to single?
// #define NO_DOUBLE_PRECISION

// Does the compiler support isnan and isinf floating point functions?
// #define NO_ISNAN_ISINF

/*************************************************************************
Formatted item width

Since it is extremely unlikely that you will ever want to use a formatted
width for a single item of more than 127 chars (i.e. the expanded and
padded size of a single % expression), the width variables can normally
be restricted to 8 bits. On small micros this saves a lot of code and
variable space. On a 32-bit RISC it may increase code size due to type
conversions. Choose the variable type to suit your CPU.
Note that a signed type is required.
*************************************************************************/

typedef signed char width_t;

/*************************************************************************
Feature configuration

This section defines the various feature flags.
These are combined as needed to produce the FEATURE_FLAGS macro.
*************************************************************************/

// Include floating point number support
#define USE_FLOAT       (1<<0)

// Include support for long integers
#define USE_LONG        (1<<1)

// Include support for octal formatting
#define USE_OCTAL       (1<<2)

// Include support for %d decimal formatting
#define USE_SIGNED      (1<<3)

// Include support for the %i synonym for %d
#define USE_SIGNED_I    (1<<4)

// Include support for the %u unsigned decimal specifier
#define USE_UNSIGNED    (1<<5)

// Include support for the %x hex specifier (lowercase output)
#define USE_HEX_LOWER   (1<<6)

// Include support for the %X hex specifier (uppercase output)
#define USE_HEX_UPPER   (1<<7)

// Force uppercase output with %x.
// Used in conjunction with USE_HEX_LOWER.
// Ignored if USE_HEX_UPPER is also set.
#define USE_HEX_UPPER_L (1<<8)

// Include support for %c single character
#define USE_CHAR        (1<<9)

// Include support for %s string
#define USE_STRING      (1<<10)

// Include support for %S string in flash memory
#define USE_FSTRING     (1<<11)

// Include support for %b binary specifier
#define USE_BINARY      (1<<12)

// Include precision support when floating point is not present.
// Precision is automatically enabled when floating point support is used.
#define USE_PRECISION   (1UL<<16)

// Allow use of leading zero padding e.g. "%03d"
#define USE_ZERO_PAD    (1UL<<17)

// Allow use of space padding e.g. "%3d" or "%12s"
#define USE_SPACE_PAD   (1UL<<18)

// Include indirect width/precision support e.g. "%*d"
#define USE_INDIRECT    (1UL<<19)

// Allow forcing a leading plus sign e.g. "%+3d"
#define USE_PLUS_SIGN   (1UL<<20)

// Allow forcing a leading space (instead of + or -) in front of zero e.g. "% 3d"
#define USE_SPACE_SIGN  (1UL<<21)

// Include support for the left-justify '-' flag.
#define USE_LEFT_JUST   (1UL<<22)

// Include support for the special '#' flag.
#define USE_SPECIAL     (1UL<<23)

/*************************************************************************
Pre-defined feature sets

This section provides some commonly used combinations of features.
*************************************************************************/

// Lowercase hex integers only.
#define HEX_INT ( \
        USE_HEX_LOWER )

// Decimal and lowercase hex only.
#define MINIMAL_INT ( \
        USE_SIGNED | USE_HEX_LOWER )

// Signed and unsigned decimal, lower case hex, zero & space padding, plus char and string
#define BASIC_INT ( \
        USE_CHAR | USE_STRING | \
        USE_SIGNED | USE_UNSIGNED | USE_HEX_LOWER | \
        USE_ZERO_PAD | USE_SPACE_PAD )

// All short integer features except octal, binary, %i, indirection and specials.
#define SHORT_INT ( \
        USE_CHAR | USE_STRING | \
        USE_SIGNED | USE_UNSIGNED | USE_HEX_LOWER | USE_HEX_UPPER | \
        USE_PRECISION  | USE_ZERO_PAD   | USE_SPACE_PAD | \
        USE_PLUS_SIGN  | USE_SPACE_SIGN | USE_LEFT_JUST )

// As above, but also supports long integers.
#define LONG_INT ( \
        USE_LONG | SHORT_INT )

// All possible integer features.
#define FULL_INT ( \
        USE_BINARY | USE_OCTAL | USE_SIGNED_I | USE_INDIRECT | USE_SPECIAL | LONG_INT )

// All available features including floating point.
#define FULL_FLOAT ( \
        USE_FLOAT | FULL_INT )

// Custom feature set. Comment out features you don't want.
#define CUSTOM_SET ( \
        USE_FLOAT       \
        USE_LONG        \
        USE_BINARY      \
        USE_OCTAL       \
        USE_SIGNED      \
        USE_SIGNED_I    \
        USE_UNSIGNED    \
        USE_HEX_LOWER   \
        USE_HEX_UPPER   \
        USE_HEX_UPPER_L \
        USE_CHAR        \
        USE_STRING      \
        USE_FSTRING     \
        USE_PRECISION   \
        USE_ZERO_PAD    \
        USE_SPACE_PAD   \
        USE_INDIRECT    \
        USE_PLUS_SIGN   \
        USE_SPACE_SIGN  \
        USE_LEFT_JUST   \
        USE_SPECIAL     \
)

// Current feature set. Use a pre-defined set or define your own.
#define FEATURE_FLAGS   SHORT_INT

#endif
